block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockMetadataStore.scala

```diff
@@ -64,8 +74,41 @@ object BlockMetadataStore {
         _ <- store.put(block.blockHash, block)
       } yield ()
 
-    def addFinalizedBlock(finalizedBlockHash: BlockHash): F[Unit] =
-      dagState.modify(st => st.copy(finalizedBlockSet = st.finalizedBlockSet + finalizedBlockHash))
+    /** Record new last finalized lock. Directly finalized is the output of finalizer,
+      * indirectly finalized are new LFB ancestors. */
+    def recordFinalized(directly: BlockHash, indirectly: Set[BlockHash])(
+        implicit sync: Sync[F]
+    ): F[Unit] = {
+      implicit val s = new Show[BlockHash] {
+        override def show(t: BlockHash): String = PrettyPrinter.buildString(t)
+      }
+
+      for {
+        // read current values
+        curMetaForDF  <- store.getUnsafe(directly)
+        curMetasForIF <- store.getUnsafeBatch(indirectly.toList)
+        // new values to persist
+        newMetaForDF  = (directly, curMetaForDF.copy(finalized = true, directlyFinalized = true))
+        newMetasForIF = curMetasForIF.map(v => (v.blockHash, v.copy(finalized = true)))
+        // update in memory state
+        _ <- dagState.modify(
+              st => {
+                val newFinalizedSet = st.finalizedBlockSet ++ indirectly + directly
+
+                // update lastFinalizedBlock only when current one is lower
+                if (st.lastFinalizedBlock.exists { case (_, h) => h > curMetaForDF.blockNum })
+                  st.copy(finalizedBlockSet = newFinalizedSet)
+                else
+                  st.copy(
+                    finalizedBlockSet = newFinalizedSet,
+                    lastFinalizedBlock = (directly, curMetaForDF.blockNum).some
+                  )
+              }
+            )
+        // persist new values all at once
+        _ <- store.put(newMetaForDF +: newMetasForIF)
```

Comment:
Connected to this comment.
https://github.com/rchain/rchain/pull/3459/files#r663200787

Between store read and write, another thread can call the same function and it's possible to overwrite others data.

I know that we don't have any other update of BlockMetaData and now is safe. But until we have "real" transactions this is just as precaution.

Reply:
Not sure wdy mean by "real" transactions?

Reply:
https://github.com/rchain/rchain/pull/3459/commits/6b5d545bb0a07ec63bbdf28aafa1080cd48fbc98